executor/common_kvm_arm64.h

```diff
@@ -63,103 +56,209 @@ static struct addr_size alloc_guest_mem(struct addr_size* free, size_t size)
+	vcpu_set_reg(cpufd, KVM_ARM64_REGS_X1, cpu_id);
+}
+
+static void install_user_code(int cpufd, void* user_text_slot, int cpu_id, const void* text, size_t text_size)
+{
+	if ((cpu_id < 0) || (cpu_id >= SYZ_KVM_MAX_VCPU))
+		return;
+	if (!user_text_slot)
+		return;
+	if (text_size > SYZ_KVM_PAGE_SIZE)
+		text_size = SYZ_KVM_PAGE_SIZE;
+	void* target = (void*)((uint64)user_text_slot + (SYZ_KVM_PAGE_SIZE * cpu_id));
+	memcpy(target, text, text_size);
+	reset_cpu_regs(cpufd, cpu_id, text_size);
+}
+
+static void setup_cpu_with_opts(int vmfd, int cpufd, const struct kvm_opt* opt, int opt_count)
+{
+	uint32 features = 0;
+	if (opt_count > 1)
+		opt_count = 1;
+	for (int i = 0; i < opt_count; i++) {
+		uint64 typ = opt[i].typ;
+		uint64 val = opt[i].val;
+		switch (typ) {
+		case 1:
+			features = val;
+			break;
+		}
+	}
 
 	struct kvm_vcpu_init init;
 	// Queries KVM for preferred CPU target type.
 	ioctl(vmfd, KVM_ARM_PREFERRED_TARGET, &init);
 	init.features[0] = features;
 	// Use the modified struct kvm_vcpu_init to initialize the virtual CPU.
 	ioctl(cpufd, KVM_ARM_VCPU_INIT, &init);
+}
 
-	// Set up CPU registers.
-	// PC points to the relative offset of guest_main() within the guest code.
-	vcpu_set_reg(cpufd, KVM_ARM64_REGS_PC, ARM64_ADDR_EXECUTOR_CODE + ((uint64)guest_main - (uint64)&__start_guest));
-	vcpu_set_reg(cpufd, KVM_ARM64_REGS_SP_EL1, ARM64_ADDR_EL1_STACK_BOTTOM + page_size - 128);
-	// Pass parameters to guest_main().
-	vcpu_set_reg(cpufd, KVM_ARM64_REGS_X0, text_size);
+#endif
 
+#if SYZ_EXECUTOR || __NR_syz_kvm_setup_cpu
+// syz_kvm_setup_cpu(fd fd_kvmvm, cpufd fd_kvmcpu, usermem vma[24], text ptr[in, array[kvm_text, 1]], ntext len[text], flags flags[kvm_setup_flags], opts ptr[in, array[kvm_setup_opt, 0:2]], nopt len[opts])
+static volatile long syz_kvm_setup_cpu(volatile long a0, volatile long a1, volatile long a2, volatile long a3, volatile long a4, volatile long a5, volatile long a6, volatile long a7)
+{
+	const int vmfd = a0;
+	const int cpufd = a1;
+	void* const host_mem = (void*)a2;
+	const struct kvm_text* const text_array_ptr = (struct kvm_text*)a3;
+	const uintptr_t text_count = a4;
+	const uintptr_t flags = a5;
+	const struct kvm_opt* const opt_array_ptr = (struct kvm_opt*)a6;
+	uintptr_t opt_count = a7;
+
+	(void)flags;
+	(void)opt_count;
+
+	(void)text_count; // fuzzer can spoof count and we need just 1 text, so ignore text_count
+	int text_type = text_array_ptr[0].typ;
+	const void* text = text_array_ptr[0].text;
+	size_t text_size = text_array_ptr[0].size;
+	(void)text_type;
+
+	void* user_text_slot = NULL;
+	setup_vm(vmfd, host_mem, &user_text_slot);
+	setup_cpu_with_opts(vmfd, cpufd, opt_array_ptr, opt_count);
+
+	install_user_code(cpufd, user_text_slot, /* cpu_id */ 0, text, text_size);
 	return 0;
 }
 #endif
 
+#if SYZ_EXECUTOR || __NR_syz_kvm_setup_syzos_vm || __NR_syz_kvm_add_vcpu
+struct kvm_syz_vm {
+	int vmfd;
+	int next_cpu_id;
+	int cpufd[SYZ_KVM_MAX_VCPU];
```

Comment:
Do we need this field? It seems to be only written to. Or will it be used later?

Reply:
Let's remove for now. Why I care is C reproducers: all unused stuff, wrappers, abstractions, etc bloat each reproducer. I usually tried to keep some balance between is best for prod code, and what's best for most readable/compact reproducers.